### DISCUSSION TITLE
For the game_latency in source code(win32)

### DISCUSSION THREAD

**Discussion by WuZemyp** [alvr-org/ALVR#1702] [Q&A]
This is not a bug, l am learning through the alvr source code and when l got confused by the variable game_latency, could l ask what is this duration? l searched in the source code and found it is the frame_present-tracking_received, but l can't find the exact frame present report point in the source code for win32. When will the report_frame_present be called? Thank you for your help!

**Comment by zmerp** [comment]
We call `report_frame_present` in `report_present`, which is passed to C++ as the function pointer `ReportPresent`, which is called on  `OvrDirectModeComponent::Present()`. Since there are a lot of jumps between Rust and C++, to search a symbol in the codebase you may need to do a global search and filter with "whole word" and "exact casing".

**Comment by WuZemyp** [comment]
> We call `report_frame_present` in `report_present`, which is passed to C++ as the function pointer `ReportPresent`, which is called on `OvrDirectModeComponent::Present()`. Since there are a lot of jumps between Rust and C++, to search a symbol in the codebase you may need to do a global search and filter with "whole word" and "exact casing".

Thank you for your reply! Acctually, l have found this point in OvrDirectModeComponent::Present(), but then l don't know where is the point that called this function. It seems that this Present() function is not called in win32. From my understanding, when the server received tracking information from client, server need to pass it to steamvr and then wait for the response of steamvr to get the frame,is the game latency mean the duration of the process of steamvr to produce the new frame? Or could you please give me a pipeline of this process. In my understanding, l found client_send_tracking->server_receive_tracking->???->frame_encode->send_to_client->client_decode->client_render. So what are the steps between server_receive_tracking and frame_encode.
Thank you in advance!

**Comment by zmerp** [comment]
Tracking data is passed to C++ code with `SetTracking()`, and then submitted to SteamVR with `vr::VRServerDriverHost()->TrackedDevicePoseUpdated()`. SteamVR will then use the pose and call `OvrDirectModeComponent::Present()` for us; `Present` overrides `IVRDriverDirectModeComponent::Present().

**Comment by WuZemyp** [comment]
Thank you for your help! So it means alvr create a virtual device called hmd by using openvr and then when server received the tracking information,it will update the tracking into the virtual device hmd. And later steamvr will also use openvr api to get the tracking information and use the virtual device to render the game frame. Then the server will get the game frame from the virtual device and use it to do encoding and then send to the physical connected device. Is that right?
Thanks!

**Comment by zmerp** [comment]
Yeah, you got everything correct 🙂
